### PR TITLE
[Feature] --substate-encoding flag to switch between rlp/others

### PR DESCRIPTION
**Pull request by rpl-ffl**
~~This change is in anticipation of the following change in substate: https://github.com/Fantom-foundation/Substate/pull/86.~~

Aida can now select between rlp/protobuf when decoding from disk using the flag `--substate-encoding`.
The flag defaults to "default" -> which defaults to rlp.

Currently effective only for `aida-vm`.

Example usage:

```
# reading rlp with rlp
$ go run ./cmd/aida-vm --validate --chainid 1 --workers 1 --evm-impl aida --vm-impl geth --aida-db ~/dev/data/rlp  --substate-encoding rlp 0 100000

...
NOTICE: Total elapsed time: 0s; last block 46147; total transaction rate ~614.20 Tx/s, ~12.90 MGas/s
```

```
# reading rlp with pb
$ go run ./cmd/aida-vm --validate --chainid 1 --workers 1 --evm-impl aida --vm-impl geth --aida-db ~/dev/data/rlp  --substate-encoding rlp 0 100000

...
cannot decode data into protobuf block: 46147, tx 0; proto: cannot parse invalid wire-format data
```

```
# reading pb with pb
$ go run ./cmd/aida-vm --validate --chainid 1 --workers 1 --evm-impl aida --vm-impl geth --aida-db ~/dev/data/0-10M --substate-encoding pb 0 100000

...
NOTICE: Total elapsed time: 0s; last block 46147; total transaction rate ~25.80 Tx/s, ~0.54 MGas/s
```

```
# reading pb with rlp
$ go run ./cmd/aida-vm --validate --chainid 1 --workers 1 --evm-impl aida --vm-impl geth --aida-db ~/dev/data/0-10M --substate-encoding rlp 0 100000

...
cannot decode data into rlp block: 46147, tx 0; rlp: expected input list for rlp.RLP
```

```
# flag unset = read with rlp, backward compatible
$ go run ./cmd/aida-vm --validate --chainid 1 --workers 1 --evm-impl aida --vm-impl geth --aida-db ~/dev/data/rlp 0 100000

...
NOTICE: Total elapsed time: 0s; last block 46147; total transaction rate ~483.26 Tx/s, ~10.15 MGas/s
```

## Type of change
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by cabrador — DISMISSED** [review]
Thanks for doing all the work on `protobuf`!

File: utils/flags.go [hunk 1]
```diff
@@ -447,4 +452,9 @@ var (
 		Usage:   "determines number of workers",
 		Value:   4,
 	}
+	UseGethBlockProcessorFlag = cli.BoolFlag{
+		Name:  "use-geth-block-processor",
+		Usage: "When set, use go-ethereum/core instead of go-opera/evmcore when ApplyMessage()",
+		Value: false,
+	}
```
Comment (cabrador) [inline utils/flags.go]: I think this flag is unrelated to this PR and is added [here](https://github.com/Fantom-foundation/Aida/pull/1171)

File: utils/config.go [hunk 2]
```diff
@@ -249,6 +250,7 @@ type Config struct {
 	UpdateDb                 string         // update-set directory
 	UpdateOnFailure          bool           // if enabled and continue-on-failure is also enabled, this updates any error found in StateDb
 	UpdateType               string         // download datatype
+	UseGethBlockProcessor    bool           // if true, uses go-ethereum/core instead of go-opera/evmcore when ApplyMessage()
```
Comment (cabrador) [inline utils/config.go]: unrelated to this PR

File: utils/config.go [hunk 3]
```diff
@@ -779,6 +781,7 @@ func (cc *configContext) reportNewConfig() {
 	log.Infof("Chain id: %v (record & run-vm only)", cfg.ChainID)
 	log.Infof("SyncPeriod length: %v", cfg.SyncPeriodLength)
 	log.Noticef("Used EVM implementation: %v", cfg.VmImpl)
+	log.Noticef("Forced geth block processor?: %v", cfg.UseGethBlockProcessor)
```
Comment (cabrador) [inline utils/config.go]: unrelated to this PR

**Review by cabrador — CHANGES_REQUESTED** [review]

File: utils/flags.go (hunk 1 above)
Comment (rpl-ffl) [reply utils/flags.go]: Thanks for the catch, removed the stray commit.

**Review by wsodsong — DISMISSED** [review]

File: utils/flags.go [hunk 4]
```diff
@@ -219,6 +219,11 @@ var (
 		Usage: "select a state DB variant to shadow the prime DB implementation",
 		Value: "",
 	}
+	SubstateEncodingFlag = cli.StringFlag{
+		Name:  "substate-encoding",
+		Usage: "select encoding when reading substate from disk: rlp (default) or protobuf",
+		Value: "default",
```
Comment (matejmode) [inline utils/flags.go]: Default value should be `rpl` or `protobuf` doesn't feel right to determine default behaviour in `Substate.db.newSubstateEncoding`

Comment (rpl-ffl) [reply utils/flags.go]: default->rlp

**Review by matejmode — DISMISSED** [review]

File: executor/substate_provider.go [hunk 5]
```diff
@@ -31,13 +33,21 @@ import (
...
 
 // OpenSubstateProvider opens a substate database as configured in the given parameters.
-func OpenSubstateProvider(cfg *utils.Config, ctxt *cli.Context, aidaDb db.BaseDB) Provider[txcontext.TxContext] {
+func OpenSubstateProvider(cfg *utils.Config, ctxt *cli.Context, aidaDb db.BaseDB) (Provider[txcontext.TxContext], error) {
 	substateDb := db.MakeDefaultSubstateDBFromBaseDB(aidaDb)
+
+	if cfg.SubstateEncoding != "" {
+		_, err := substateDb.SetSubstateEncoding(cfg.SubstateEncoding)
+		if err != nil {
+			return nil, fmt.Errorf("Fail to set substate encoding; %w", err)
```
Comment (cabrador) [inline executor/substate_provider.go]: errors in Go should be lower case

File: executor/substate_provider.go [hunk 6]
```diff
@@ -31,13 +33,21 @@ import (
 // ----------------------------------------------------------------------------
 
 // OpenSubstateProvider opens a substate database as configured in the given parameters.
-func OpenSubstateProvider(cfg *utils.Config, ctxt *cli.Context, aidaDb db.BaseDB) Provider[txcontext.TxContext] {
+func OpenSubstateProvider(cfg *utils.Config, ctxt *cli.Context, aidaDb db.BaseDB) (Provider[txcontext.TxContext], error) {
 	substateDb := db.MakeDefaultSubstateDBFromBaseDB(aidaDb)
+
+	if cfg.SubstateEncoding != "" {
```
Comment (cabrador) [inline executor/substate_provider.go:39]: Why do we need to check it? I thought `""` would chose a default option.

**Review by cabrador — DISMISSED** [review]
Only minor comments. Otherwise LGTM!

File: executor/substate_provider.go (hunk 6 above)
Comment (rpl-ffl) [reply executor/substate_provider.go:39]: `cfg.SubstateEncoding` was set to `"default"` by the flags/config system, so manually created config in tests would yield `""`. This was the fastest way to get this version of Substate to work.

Changes where `""` also defensively defaults to `rlp` will be in a separate PR.

File: executor/substate_provider.go (hunk 5 above)
Comment (rpl-ffl) [reply executor/substate_provider.go]: Fixed
[842a212](https://github.com/Fantom-foundation/Aida/pull/1170/commits/842a212a3be4247eb37cd7f1b0615087e160fedc)

**Review by cabrador — APPROVED** [review]
